src/endpoints/http/collections/export/lib/parser.js

```diff
@@ -0,0 +1,40 @@
+import stream from 'stream';
+import jcsv from 'jcsv';
+import concat from 'concat-stream';
+import str from 'string-to-stream';
+
+class json extends stream.Transform {
+  constructor() {
+    super({ objectMode: true });
+  }
+
+  _transform(data, encoding, cb) {
+    data = JSON.stringify(data);
+    this.push(data);
+    cb();
+  }
+}
+
+export default {
+  bson: collection => Promise.resolve(collection),
+  json: collection => {
+    const parserJson = new json();
+    parserJson.filename = collection.filename;
+    return Promise.resolve(collection.pipe(parserJson));
+  },
+  csv: collection => {
+    const promise = new Promise((resolve, reject) => {
+      collection.pipe(concat(jsonArray => {
+        jcsv(jsonArray, { separator: ',', newline: "\r\n", headers: true }, (err, csv) => {
```

Comment:
@wei-lee The issue here is that for CSV we need to specify the headers before we start piping as the headers are the first line in the CSV file. And we won't know the headers until we load all the data and check for all the unique properties on all docs in the collection.
`jcsv`  generates the headers for us and we pipe it from that point on to keep the stream interface across all formats.

The `csv-write-stream` docs says that `if you don't specify a header row the keys of the first row written to the stream will be used as the header row` but we can't guarantee that all docs have the same properties.

Do you know a more efficient way to get the headers (list of all unique properties on docs in the collection) ?